Firestore/Source/Remote/FSTDatastore.m

```diff
@@ -557,21 +609,79 @@ - (void)resumeStartFromBackoff {
 
   // Momentarily set state to FSTStreamStateInitial as `start` expects it.
   self.state = FSTStreamStateInitial;
-  [self start];
+  [self start:delegate];
   FSTAssert([self isStarted], @"Stream should have started.");
 }
 
-- (void)stop {
-  FSTLog(@"%@ %p stop", NSStringFromClass([self class]), (__bridge void *)self);
+/**
+ * Closes the stream and cleans up as necessary:
+ *
+ * <ul>
+ *   <li>closes the underlying GRPC stream;
+ *   <li>calls the onClose handler with the given 'error';
+ *   <li>sets internal stream state to 'finalState';
+ *   <li>adjusts the backoff timer based on the error
+ * </ul>
+ *
+ * A new stream can be opened by calling `start` unless `finalState` is set to
+ * `FSTStreamStateStopped`.
+ *
+ * @param finalState the intended state of the stream after closing.
+ * @param error the NSError the connection was closed with.
+ */
+- (void)close:(FSTStreamState)finalState error:(NSError *_Nullable)error {
+  FSTAssert(finalState == FSTStreamStateError || error == nil,
+            @"Can't provide an error when not in an error state.");
+
   [self.workerDispatchQueue verifyIsCurrentQueue];
+  [self cancelIdleCheck];
 
-  // Prevent any possible future restart of this stream.
-  self.state = FSTStreamStateStopped;
+  if (finalState != FSTStreamStateError) {
+    // If this is an intentional close ensure we don't delay our next connection attempt.
+    [self.backoff reset];
+  } else if (error != nil && error.code == FIRFirestoreErrorCodeResourceExhausted) {
+    FSTLog(@"%@ %p Using maximum backoff delay to prevent overloading the backend.", [self class],
+           (__bridge void *)self);
+    [self.backoff resetToMax];
+  }
 
-  // Close the stream client side.
-  FSTBufferedWriter *requestsWriter = self.requestsWriter;
-  @synchronized(requestsWriter) {
-    [requestsWriter finishWithError:nil];
+  // This state must be assigned before calling `notifyStreamInterrupted` to allow the callback to
+  // inhibit backoff or otherwise manipulate the state in its non-started state.
+  self.state = finalState;
+
+  if (self.requestsWriter) {
+    // Clean up the underlying RPC. If this close: is in response to an error, don't attempt to
+    // call half-close to avoid secondary failures.
+    if (finalState != FSTStreamStateError) {
+      FSTLog(@"%@ %p Closing stream client-side", [self class], (__bridge void *)self);
+      @synchronized(self.requestsWriter) {
+        [self.requestsWriter finishWithError:nil];
+      }
+    }
+    _requestsWriter = nil;
+  }
+
+  // If the caller explicitly requested a stream stop, don't notify them of a closing stream (it
+  // could trigger undesirable recovery logic, etc.).
+  if (finalState != FSTStreamStateStopped) {
```

Comment:
We should avoid sending messages to delegates in the middle of operations like these because they may attempt to perform operations on the stream as a result of the callback. If that happened here the stream would be in an inconsistent state.

Reply:
I moved some of the clean up logic above. I am not able to move the clearing of the delegate out, since it needs to be assigned in notifyStreamInterruptedWithError.